### DISCUSSION TITLE
Why the cookies are not blocked when the "opt-in" option is turned on?

### DISCUSSION THREAD

**Discussion by rodrigo-rufino** [orestbida/cookieconsent#491] [General]
Only the scripts are blocked in this case. To guarantee that the cookies are compliant with the user consent, shouldn't they also be blocked together with the scripts?

What if I know that a cookie exists, but I don't know what script is setting them? How would I block this specific cookie until the user consents to it?

**Comment by orestbida** [comment]
The plugin does not "block" cookies, it manages the scripts which set cookies. `mode: 'opt-in'` simply means that the configured scripts will run after the user's explicit consent.

> What if I know that a cookie exists, but I don't know what script is setting them?

Unfortunately the only way is to analyze the website's scripts manually. If you have a local copy of the website, you can use [vscode](https://code.visualstudio.com/) to do a global search for the `document.cookie` keyword and see what pops up. You can also use [Open Cookie Database](https://github.com/jkwakman/Open-Cookie-Database/blob/master/open-cookie-database.csv) to know what a specific cookie does.

**Comment by rodrigo-rufino** [comment]
Hmm, but the "_autoclearCookies()" method does exactly that, doesn't it? It clears the cookie when the user rejects the non-essential categories.

So, shouldn't the cookies also be cleared when the mode is set to `'opt-in'`?

**Reply by orestbida** [reply]
the autoclear function only works if a valid cookie table is configured. When the user rejects a specific category and there is a valid cookie table, then the autoclear function will try to erase the cookies specified in that table.

There is no "automatic cookie detection", that's just not possible. You'd need some kind of AI based tool that can scan your entire website to determine with a decent degree of accuracy what cookies are being set and what they do.

**Reply by rodrigo-rufino** [reply]
But the point is: why doesn't the autoclear function run right after the Cookie Consent component loads? Why does it only run after the user rejects the non-essential categories?

**Reply by orestbida** [reply]
The autoclear function runs the very first time consent is expressed (accept/reject action) and whenever a category is disabled (after the very first consent). 

Why would you want it to run every time? If cookies are being re-set, then it means that there are scripts that are not being handled by the plugin. Erasing cookies does not make you compliant, disabling the scripts that set them (generally tracking scripts) does. 

If you really want to erase these cookies, you can use the `.eraseCookies` method inside the `onAccept` callback.

**Reply by rodrigo-rufino** [reply]
> Why would you want it to run every time?

Not exactly every time. I just wanted to understand why the cookies were not erased when the scripts were blocked.

The current behavior is:

1. Cookie Consent component is loaded with `mode = 'opt-in'`.
2. Scripts are blocked.
3. User clicks on Reject.
4. Scripts keep being blocked, **and non-essential cookies are erased**.

But shouldn't it behave like this?
1. Cookie Consent component is loaded with `mode = 'opt-in'`.
2. Scripts are blocked. **Previously set non-essential cookies are erased**.
3. User clicks on Reject.
4. Scripts keep being blocked.


For me, if the cookies are cleared when the user withdraws consent (clicks on Reject), they also should be cleared when the user didn't give consent yet (didn't click on accept and `mode = 'opt-in'`).

**Reply by orestbida** [reply]
> ... they also should be cleared when the user didn't give consent yet (didn't click on accept and mode = 'opt-in').

This occurs at the very first consent. 

1. Reject all categories except necessary ones (very first consent)  => no scripts enabled => all (non-necessary) cookies are erased via the autoclear function.

2. User reloads the page => no scripts enabled => no cookies to clear => no need to re-run the autoclear function.

If after a page reload, a cookie — belonging to one of the rejected categories — is set, it means that the plugin is not properly configured to block scripts.

This is the intended/desired behavior. I'm struggling to understand how this impacts your setup.

**Comment by benseven** [comment]
But the non-essential cookies you’re referring to won’t have been set yet
because your script is blocked until the user accepts? So they don’t exist
and as such don’t need cleared. If you have scripts setting cookies before
acceptance then that’s the problem, they should be blocked until consent.

On Mon, 20 Mar 2023 at 17:47, Rodrigo Rufino Ribeiro <
***@***.***> wrote:

> Why would you want it to run every time?
>
> Not exactly every time. I just wanted to understand why the cookies were
> not erased when the scripts were blocked.
>
> The current behavior is:
>
>    1. Cookie Consent component is loaded with mode = 'opt-in'.
>    2. Scripts are blocked.
>    3. User clicks on Reject.
>    4. Scripts keep being blocked, *and non-essential cookies are erased*.
>
> But shouldn't it behave like this?
>
>    1. Cookie Consent component is loaded with mode = 'opt-in'.
>    2. Scripts are blocked. *Previously set non-essential cookies are
>    erased*.
>    3. User clicks on Reject.
>    4. Scripts keep being blocked.
>
> For me, if the cookies are cleared when the user withdraws consent (clicks
> on Reject), they also should be cleared when the user didn't give consent
> yet (didn't click on accept and mode = 'opt-in').
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/orestbida/cookieconsent/discussions/491#discussioncomment-5372619>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AANAZC2LGM34GFZFJJOSKJTW5CJ2JANCNFSM6AAAAAAWBBGIYU>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***
> com>
>
-- 
ben seven is a designer and web developer, the co-founder of
http://kin.works and founder of https://devicelab.org
*he frequently has to say 'no, sorry' to work or favours to focus on family
and existing projects, and is grateful for your understanding.*
+44 7746 200412 | ***@***.***

**Reply by rodrigo-rufino** [reply]
Not if the Cookie Consent component is integrated to an already existing domain. Cookies added before the component integration would still be there.

**Reply by JonL1** [reply]
If some cookies were already set before you could ask for consent it's not really opt-in, it's opt-out.

**Comment by benseven** [comment]
> Not if the Cookie Consent component is integrated to an already existing
domain.

I don’t think it’s quite right to expect a library to cater to the specific
implementation details of a pre-existing site.

If I was in this position I’d be clearing or overwriting the existing
cookies set by the domain with a piece of custom code, and instantiating
cookie-consent after that. But I wouldn’t expect the library to handle it,
because only I as the author know the names and roles of those cookies and
they’ve never been handled by the library in the first place.

On Mon, 20 Mar 2023 at 18:10, Rodrigo Rufino Ribeiro <
***@***.***> wrote:

> Not if the Cookie Consent component is integrated to an already existing
> domain. Cookies added before the component integration would still be there.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/orestbida/cookieconsent/discussions/491#discussioncomment-5372861>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AANAZC45UM4TEYTUJSX6QILW5CMS5ANCNFSM6AAAAAAWBBGIYU>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>
-- 
ben seven is a designer and web developer, the co-founder of
http://kin.works and founder of https://devicelab.org
*he frequently has to say 'no, sorry' to work or favours to focus on family
and existing projects, and is grateful for your understanding.*
+44 7746 200412 | ***@***.***
